### PR TITLE
Refactor DeepSeek MoE layer to unify the two forward branches

### PR DESCRIPTION
**Pull request by fzyzcjy**
## Motivation

With this, it is easier to extract everything into ops for a unified TBO.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -299,40 +300,34 @@ def __init__(
...
-            shared_output = self._forward_shared_experts(hidden_states)
+        else:
+            router_logits = None
+
+        if (
+            (self.n_share_experts_fusion == 0)
+            and (forward_mode is not None)
+            and not forward_mode.is_idle()
+            and hidden_states.shape[0] > 0
+        ):
```
Comment (yizhang2077) [inline python/sglang/srt/models/deepseek_v2.py:316]: I think here should be `(self.n_share_experts_fusion == 0) and ( ( not self._enable_deepep_moe ) or blabla)`? 
And condition code seems have 3 lines the same as above place, it is better to have a helper function?

File: python/sglang/srt/models/deepseek_v2.py [hunk 2]
```diff
@@ -299,40 +300,34 @@ def __init__(
...
+            shared_output = self.self.shared_experts(hidden_states)
```
Comment (yizhang2077) [inline python/sglang/srt/models/deepseek_v2.py]: typo?

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (yizhang2077) [reply python/sglang/srt/models/deepseek_v2.py:316]: Btw, it is extracted here for future TBO?

**Review by yizhang2077 — COMMENTED** [review]
There are some comments need to be address. Apart from that, LGTM.

File: python/sglang/srt/models/deepseek_v2.py (hunk 2 above)
Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py]: yes typo (fixed in the merged PR but forgot this one...)

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py:316]: > Btw, it is extracted here for future TBO?

yes

Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py:316]: > And condition code seems have 3 lines the same as above place, it is better to have a helper function?

I will do it in the last PR in chain

EDIT: originally thought this pattern occurs many times, but now check and only occur here. Then wondering whether we need to (a) use this pattern everywhere, or (b) remove this pattern here. Anyway, I personally think we can do it later since it is a separate issue.

Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py:316]: > not self._enable_deepep_moe

looks reasonable, will change it

EDIT: ~~hmm, when in not-deepep case, if the hidden states are empty or idle, it seems good to skip computation as well~~ changed, you are correct, when in DP attention mode without deepep and one rank is idle, we still need to compute other ranks' data since it is TP in mlp

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -194,6 +194,14 @@ def forward(self, hidden_states):
         return logits
 
 
+def is_non_idle_and_non_empty(forward_mode, hidden_states):
+    return (
+        (forward_mode is not None)
```
Comment (yizhang2077) [inline python/sglang/srt/models/deepseek_v2.py:199]: I think it can be moved to a common file, since other moe models also can use it, but it can be moved in later pr

Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py:199]: I also thought about this, but realize no other files checked this... e.g. see below, I searched "is_idle", and all other usages are only checking is_idle and not check shape[0] == 0

image
image

**Review by yizhang2077 — APPROVED** [review]

File: python/sglang/srt/models/deepseek_v2.py (hunk 3 above)
Comment (yizhang2077) [reply python/sglang/srt/models/deepseek_v2.py:199]: I think qwen3 will need this? We keep it here now, and move it into common file if qwen3 need it

Comment (fzyzcjy) [reply python/sglang/srt/models/deepseek_v2.py:199]: Sure, looks reasonable
